### PR TITLE
WIP Update documentating guide

### LINKED ISSUE #9786: Consistent Documentation Guide for Docstrings
### Bug report

**Summary**

The [documentation guide](http://matplotlib.org/devdocs/devel/documenting_mpl.html#documenting-matplotlib) contains a section about docstrings. Unfortunately, in my attempt to add / improve some docstrings I found it largely deficient.

This issue collects the problems I encounted and may serve as a starting point to improve the API docs.

**Motivation**

A good documentation is essentially for the usability of a library. Even more so, for a large library with many options and making heavy use of `*args` and `**kwargs`.

To get improvements here, there has to be a *consistent and easily understandable style guide*. Otherwise, contributors will not know or bother.

**Issues**

1. **Multiple sources**

* the  [documentation guide](http://matplotlib.org/devdocs/devel/documenting_mpl.html#documenting-matplotlib)
* [Numpy: How to document](https://github.com/numpy/numpy/blob/master/doc/HOWTO_DOCUMENT.rst.txt)
* The pending #9513 was pointed out to me as an additional source, however I don't think it says anything about docstrings.
* There are some rules discussed or decided in issues, e.g. #8079 on quotes, #7535,  #6924 on Other Parameters

- [ ] Include all information in the [docstrings section](http://matplotlib.org/devdocs/devel/documenting_mpl.html#docstrings), reference numpydoc docmentation.

2. **Outdated directions**

The [docstrings section](http://matplotlib.org/devdocs/devel/documenting_mpl.html#docstrings) is outdated and thus misleading.

> In addition to the “narrative” documentation described above, matplotlib also defines its API reference documentation in docstrings. For the most part, these are standard Python docstrings, but matplotlib also includes some features to better support documenting getters and setters.

better:

> In addition to the “narrative” documentation described above, the API reference documentation is generated from docstrings in the code. Generally, matplotlib uses numpy style docstrings [link]. To further enhance the documentation we have augmented numpy style by some additional conventions and features.
> 
> Additional Conventions:
> 
> Additional Features: [existing stuff here]

- [ ] Remove outdated stuff. Add placeholders and notes if there are some known gaps.

3. **Additional Conventions**

numpydoc does not fully cover all aspects of docstrings. Some should be clarified on the project level.

- Handling of `*args` and `**kwargs`: Should they be Parameters, Other Parameters or something else?

- Handling of long lists of value choices:

[See here](https://github.com/matplotlib/matplotlib/pull/9777#pullrequestreview-76296627):
```
projection :
            ['aitoff' | 'hammer' | 'lambert' | 'mollweide' | \
'polar' | 'rectilinear'], optional
```

- When to use *abc*, `abc`, 'abc', "abc", ``abc``. In particular also usage of double or single quotes in the docs (even if it's: Do as you like but don't change existing stuff). Indicate if these decisions deviate intentionally from other sources you link (e.g. [numpy](https://github.com/numpy/numpy/blob/master/doc/HOWTO_DOCUMENT.rst.txt): "Variable, module, function, and class names should be written between single back-ticks (`numpy`)", vs [Formatting](http://matplotlib.org/devdocs/devel/documenting_mpl.html#formatting): "Function arguments and keywords should be referred to using the emphasis role."

- When to cross reference and when not  "Types don't need to be marked up." vs "Sometimes sphinx doesn't automatically put in links with out the markup though." [See here](https://github.com/matplotlib/matplotlib/pull/9777#pullrequestreview-76296627) Also if sphinx has a bug that return values are not always cross-referenced, mention that and your policy (Either keep plain types, it will be fine once sphinx is fixed. Or add reference now, we can strip it later).

- Usage of call signature: Some methods have this, others not.

~~~
    Call signature::

      text(x, y, s, fontdict=None, **kwargs)
~~~

4. **Migration of documentation**

- [ ] State that there's still a mixture of old and new styles and that the migration is ongoing.
- [ ] List deprecated conventions and how to replace them.

5. **Examples**

- [ ] Add examples in the above parts as much as possible. They are more instructure than lengthy texts.
- [ ] Link to parts of code with good documentation, from which one can learn.

### PR DESCRIPTION
**Pull request by dstansby**
Needs a bit more work, but the aim is to fix #9786 eventually and generally clean the writing documentation guide. Comments very welcome and encouraged!

### CODE REVIEW

File: doc/devel/documenting_mpl.rst [hunk 1]
```diff
@@ -181,135 +178,81 @@ Formatting
...
+Footnotes
+---------
+Footnotes [#]_ can be added using ``[#]_``, followed later by::
 
     .. rubric:: Footnotes
 
     .. [#]
 
-  .. rubric:: Footnotes
-
```
Comment (dstansby) [inline doc/devel/documenting_mpl.rst]: I've been fairly ruthless deleting stuff here, as how to do all this properly is documented in the numpydoc guide (which I will add a link to in this document at some point)

**Comment by anntzer** [comment]
I would really appreciate if #9311 could go in first...

I had closed it because I was hoping for #9513 to be merged first and resolve the conflicts in #9311 after that, but I can resolve the conflicts in the other direction to if that's needed :/.

**Comment by jklymak** [comment]
Ah, @anntzer  thats what you were talking about the other day.  I looked at your open issues and didn't see what you meant!  I'll try and review and merge #9311 after the tests build (if someone doesn't beat me to it).

**Comment by anntzer** [comment]
I was planning to reopen it after #9513 goes it, but whatever.

**Comment by dstansby** [comment]
Woops, sorry, I forgot about those. Let me close this for now and I'll have a look at your PRs.

**Pull request closed without merging**